### PR TITLE
Optimizer API

### PR DESCRIPTION
**Pull request by Ubospica**
This pr provides interfaces for creating optimizers and integrate them into the training workflow.

Now these optimizers are supported:
- SGD (with weight decay)
- MomentumSGD (SGD with momentum, dampening, weight decay and nesterov momentum)
- Adam (not implemented)

Besides, this PR:
- Adds essential comments on `SimpleAD` and `gradient`
- Integrates builtin `OperatorLegalizer`
- Move optimizer.py, leagalizer_update.py, gradient.py under package `tvm.relax.training`
- Adds empty lines and removes useless comments
- By importing transform, we now automatically imports `tvm.relax.training.gradient`. This is because SimpleAD requires gradient ops.

Known issues:
- OperatorLegalizer calls RemoveUnusedVars, which does not work well with our test cases currently.

### CODE REVIEW

**Review by SiriusNEO — COMMENTED** [review]
reviewed
